lib/Conversion/CombToSMT/CombToSMT.cpp

```diff
@@ -0,0 +1,245 @@
+      rewriter.replaceOpWithNewOp<smt::EqOp>(op, adaptor.getLhs(),
+                                             adaptor.getRhs());
+      return success();
+    }
+
+    if (adaptor.getPredicate() == ICmpPredicate::ne) {
+      rewriter.replaceOpWithNewOp<smt::DistinctOp>(op, adaptor.getLhs(),
+                                                   adaptor.getRhs());
+      return success();
+    }
+
+    smt::BVCmpPredicate pred;
+    switch (adaptor.getPredicate()) {
+    case ICmpPredicate::sge:
+      pred = smt::BVCmpPredicate::sge;
+      break;
+    case ICmpPredicate::sgt:
+      pred = smt::BVCmpPredicate::sgt;
+      break;
+    case ICmpPredicate::sle:
+      pred = smt::BVCmpPredicate::sle;
+      break;
+    case ICmpPredicate::slt:
+      pred = smt::BVCmpPredicate::slt;
+      break;
+    case ICmpPredicate::uge:
+      pred = smt::BVCmpPredicate::uge;
+      break;
+    case ICmpPredicate::ugt:
+      pred = smt::BVCmpPredicate::ugt;
+      break;
+    case ICmpPredicate::ule:
+      pred = smt::BVCmpPredicate::ule;
+      break;
+    case ICmpPredicate::ult:
+      pred = smt::BVCmpPredicate::ult;
+      break;
+    default:
+      llvm_unreachable("all cases handled above");
+    }
+
+    rewriter.replaceOpWithNewOp<smt::BVCmpOp>(op, pred, adaptor.getLhs(),
+                                              adaptor.getRhs());
+    return success();
+  }
+};
+
+/// Lower a comb::ExtractOp operation to an smt::ExtractOp
+struct ExtractOpConversion : OpConversionPattern<ExtractOp> {
+  using OpConversionPattern<ExtractOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(ExtractOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    rewriter.replaceOpWithNewOp<smt::ExtractOp>(
+        op, typeConverter->convertType(op.getResult().getType()),
+        adaptor.getLowBitAttr(), adaptor.getInput());
+    return success();
+  }
+};
+
+/// Lower a comb::MuxOp operation to an smt::IteOp
+struct MuxOpConversion : OpConversionPattern<MuxOp> {
+  using OpConversionPattern<MuxOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(MuxOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    Value condition = typeConverter->materializeTargetConversion(
+        rewriter, op.getLoc(), smt::BoolType::get(getContext()),
+        adaptor.getCond());
+    rewriter.replaceOpWithNewOp<smt::IteOp>(
+        op, condition, adaptor.getTrueValue(), adaptor.getFalseValue());
+    return success();
+  }
+};
+
+/// Lower a comb::SubOp operation to an smt::BVNegOp + smt::BVAddOp
+struct SubOpConversion : OpConversionPattern<SubOp> {
+  using OpConversionPattern<SubOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(SubOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    Value negRhs = rewriter.create<smt::BVNegOp>(op.getLoc(), adaptor.getRhs());
```

Comment:
We don't have a BVSub operation. We can add one, but having both `sub` and `neg` is probably not worth it? I chose `neg` because it's part of the theory, sub is only a syntactic sugar op in the QF_BV logic.